File: src/MobBuilder.java
package net.glitchtechs.starapi;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.entity.Ageable;
import org.bukkit.entity.Entity;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.LivingEntity;
import org.bukkit.inventory.EntityEquipment;
import org.bukkit.inventory.ItemStack;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;

public class MobBuilder {
    private EntityType type;
    private Location location;
    private String customName;
    private double health;
    private List<ItemStack> armorContents = new ArrayList();
    private ItemStack heldItem;
    private List<PotionEffect> potionEffects = new ArrayList();
    private boolean showParticles = true;
    private boolean isBaby;

    private MobBuilder() {
    }

    public static MobBuilder create() {
        return new MobBuilder();
    }

    public MobBuilder type(EntityType type) {
        this.type = type;
        return this;
    }

    public MobBuilder location(Location location) {
        this.location = location;
        return this;
    }

    public MobBuilder customName(String customName) {
        this.customName = MessageUtils.hex(customName);
        return this;
    }

    public MobBuilder health(double health) {
        return this;
    }

    public MobBuilder armorContents(Material... materials) {
        Material[] var2 = materials;
        int var3 = materials.length;

        for(int var4 = 0; var4 < var3; ++var4) {
            Material material = var2[var4];
            this.armorContents.add(new ItemStack(material));
        }

        return this;
    }

    public MobBuilder heldItem(Material material) {
        this.heldItem = new ItemStack(material);
        return this;
    }

    public MobBuilder showParticles(boolean showParticles) {
        this.showParticles = showParticles;
        return this;
    }

    public MobBuilder potionEffect(PotionEffectType type, int durationSeconds, int amplifier) {
        PotionEffect potionEffect = new PotionEffect(type, durationSeconds * 20, amplifier, false, this.showParticles);
        this.potionEffects.add(potionEffect);
        return this;
    }

    public MobBuilder isBaby(boolean isBaby) {
        this.isBaby = isBaby;
        return this;
    }

    public LivingEntity spawn() {
        this.validate();
        Entity spawnedEntity = this.location.getWorld().spawnEntity(this.location, this.type);
        if (!(spawnedEntity instanceof LivingEntity)) {
            throw new IllegalStateException("Spawned entity is not a LivingEntity");
        } else {
            LivingEntity mob = (LivingEntity)spawnedEntity;
            if (this.customName != null) {
                mob.setCustomName(this.customName);
                mob.setCustomNameVisible(true);
            }

            if (this.health > 0.0D) {
                mob.setHealth(this.health);
            }

            EntityEquipment equipment = mob.getEquipment();
            if (equipment != null) {
                int i = 0;

                for(Iterator var5 = this.armorContents.iterator(); var5.hasNext(); ++i) {
                    ItemStack armor = (ItemStack)var5.next();
                    equipment.setArmorContents((ItemStack[])this.armorContents.toArray(new ItemStack[0]));
                }
            }

            if (this.heldItem != null) {
                equipment.setItemInMainHand(this.heldItem);
            }

            Iterator var7 = this.potionEffects.iterator();

            while(var7.hasNext()) {
                PotionEffect effect = (PotionEffect)var7.next();
                mob.addPotionEffect(effect);
            }

            if (mob instanceof Ageable) {
                Ageable ageable = (Ageable)mob;
                ageable.setAdult();
                if (this.isBaby) {
                    ageable.setBaby();
                }
            }

            return mob;
        }
    }

    private void validate() {
        if (this.type == null || this.location == null) {
            throw new IllegalStateException("Type and location must be set before spawning the mob.");
        }
    }
}

File: Kotlin/src/MobBuilder.kt
package net.glitchtechs.starapi

import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.entity.Ageable
import org.bukkit.entity.EntityType
import org.bukkit.entity.LivingEntity
import org.bukkit.inventory.ItemStack
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType


class MobBuilder private constructor() {
    private var type: EntityType? = null
    private var location: Location? = null
    private var customName: String? = null
    private val health = 0.0
    private val armorContents: MutableList<ItemStack?> = ArrayList()
    private var heldItem: ItemStack? = null
    private val potionEffects: MutableList<PotionEffect?> = ArrayList()
    private var showParticles = true
    private var isBaby = false

    fun type(type: EntityType?): MobBuilder {
        this.type = type
        return this
    }

    fun location(location: Location?): MobBuilder {
        this.location = location
        return this
    }

    fun customName(customName: String?): MobBuilder {
        this.customName = MessageUtils.hex(customName!!)
        return this
    }

    fun health(health: Double): MobBuilder {
        return this
    }

    fun armorContents(vararg materials: Material): MobBuilder {
        val var2 = materials
        val var3 = materials.size

        for (var4 in 0 until var3) {
            val material = var2[var4]
            armorContents.add(ItemStack(material))
        }

        return this
    }

    fun heldItem(material: Material?): MobBuilder {
        this.heldItem = ItemStack(material!!)
        return this
    }

    fun showParticles(showParticles: Boolean): MobBuilder {
        this.showParticles = showParticles
        return this
    }

    fun potionEffect(type: PotionEffectType?, durationSeconds: Int, amplifier: Int): MobBuilder {
        val potionEffect = PotionEffect(type!!, durationSeconds * 20, amplifier, false, this.showParticles)
        potionEffects.add(potionEffect)
        return this
    }

    fun isBaby(isBaby: Boolean): MobBuilder {
        this.isBaby = isBaby
        return this
    }

    fun spawn(): LivingEntity {
        this.validate()
        val spawnedEntity = location!!.world!!.spawnEntity(location!!, type!!)
        check(spawnedEntity is LivingEntity) { "Spawned entity is not a LivingEntity" }
        val mob = spawnedEntity
        if (this.customName != null) {
            mob.customName = customName
            mob.isCustomNameVisible = true
        }

        if (this.health > 0.0) {
            mob.health = health
        }

        val equipment = mob.equipment
        if (equipment != null) {
            var i = 0

            val var5: Iterator<*> = armorContents.iterator()
            while (var5.hasNext()) {
                val armor = var5.next() as ItemStack
                equipment.armorContents = (armorContents.toTypedArray<ItemStack?>() as Array<ItemStack?>?)!!
                ++i
            }
        }

        if (this.heldItem != null) {
            equipment!!.setItemInMainHand(this.heldItem)
        }

        val var7: Iterator<*> = potionEffects.iterator()

        while (var7.hasNext()) {
            val effect = var7.next() as PotionEffect
            mob.addPotionEffect(effect)
        }

        if (mob is Ageable) {
            val ageable = mob
            ageable.setAdult()
            if (this.isBaby) {
                ageable.setBaby()
            }
        }

        return mob
    }

    private fun validate() {
        if (this.type == null || this.location == null) {
            throw IllegalStateException("Type and location must be set before spawning the mob.")
        }
    }

    companion object {
        fun create(): MobBuilder {
            return MobBuilder()
        }
    }
}